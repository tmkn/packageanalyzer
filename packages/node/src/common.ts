import * as path from "path";
import * as fs from "fs";
import { fileURLToPath } from "url";

import dayjs from "dayjs";

import type { DependencyTypes } from "../../shared/src/reports/Validation.js";

export const defaultDependencyType: DependencyTypes = "dependencies";

export function getVersion(): string {
    try {
        const __dirname = path.dirname(fileURLToPath(import.meta.url));
        // adjust when moving this function to another folder
        const file = path.join(__dirname, "..", "..", "..", "package.json");
        return JSON.parse(fs.readFileSync(file, "utf8")).version;
    } catch {
        return "version parse error!";
    }
}

export function daysAgo(date: string | number | Date): string {
    return `(${dayjs(new Date()).diff(date, "day")} days ago)`;
}
