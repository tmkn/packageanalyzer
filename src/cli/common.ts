import * as path from "path";
import * as fs from "fs";

import * as dayjs from "dayjs";
import * as chalk from "chalk";

import { PackageAnalytics, VersionSummary, GroupedLicenseSummary } from "../analyzers/package";
import { DependencyTypes } from "../visitors/visitor";

export const defaultDependencyType: DependencyTypes = "dependencies";

export function getVersion(): string {
    try {
        const file = path.join(__dirname, "./../../package.json");

        return JSON.parse(fs.readFileSync(file, "utf8")).version;
    } catch (e) {
        return "version parse error!";
    }
}

export function daysAgo(date: string | number | Date): string {
    return `(${dayjs(new Date()).diff(date, "day")} days ago)`;
}

export function printStatistics(pa: PackageAnalytics): void {
    const padding = 40;
    const paddingLeft = 4;

    console.log(chalk.bold(`Statistics for ${pa.fullName}\n`));

    printPublished(pa, padding);
    printOldest(pa.oldest, padding);
    printNewest(pa.newest, padding);
    console.log(`${`Direct dependencies:`.padEnd(padding)}${pa.directDependencyCount}`);
    console.log(`${`Transitive dependencies:`.padEnd(padding)}${pa.transitiveDependenciesCount}`);
    printDistinctDependencies(pa.distinctByNameCount, pa.distinctByVersionCount, paddingLeft);
    printMostReferred(pa.mostReferred, padding);
    printMostDependencies(pa.mostDependencies, padding);
    printMostVersion(pa.mostVersions, padding, paddingLeft);
    printLoops(pa, padding, paddingLeft);
    printLicenseInfo(pa.licensesByGroup, paddingLeft);
}

function printNewest(newest: PackageAnalytics | undefined, padding: number): void {
    if (newest && newest.published) {
        console.log(
            `${`Newest package:`.padEnd(padding)}${
                newest.fullName
            } - ${newest.published.toUTCString()} ${daysAgo(newest.published)}`
        );
        console.log(`${`Newest package path:`.padEnd(padding)}${newest.pathString}`);
    }
}

function printOldest(oldest: PackageAnalytics | undefined, padding: number): void {
    if (oldest && oldest.published) {
        console.log(
            `${`Oldest package:`.padEnd(padding)}${
                oldest.fullName
            } - ${oldest.published.toUTCString()} ${daysAgo(oldest.published)}`
        );
        console.log(`${`Oldest package path:`.padEnd(padding)}${oldest.pathString}`);
    }
}

function printPublished(pa: PackageAnalytics, padding: number): void {
    if (!pa.published) return;

    console.log(
        `${`Published:`.padEnd(padding)}${pa.published.toUTCString()} ${daysAgo(pa.published)}`
    );
}

function printDistinctDependencies(
    byName: number,
    byNameAndVersion: number,
    paddingLeft: number
): void {
    console.log(`Distinct dependencies:`);
    console.log(`${``.padStart(paddingLeft)}${byName}: distinct name`);
    console.log(`${``.padStart(paddingLeft)}${byNameAndVersion}: distinct name and version`);
}

function printLoops(pa: PackageAnalytics, padding: number, paddingLeft: number): void {
    const { loops, loopPathMap, distinctLoopCount } = pa;

    console.log(`${`Loops:`.padEnd(padding)}${loops.length} (${distinctLoopCount} distinct)`);

    if (distinctLoopCount > 0) {
        const [first] = loops.map(l => l.pathString).sort();

        console.log(`    affected Packages: [${[...loopPathMap.keys()].join(", ")}]`);
        console.log(`    e.g. ${first}`);

        if (loops.length > 1)
            console.log(`${``.padStart(paddingLeft)}+${loops.length - 1} additional loops`);
    }
}

function printMostDependencies(pa: PackageAnalytics, padding: number): void {
    console.log(
        `${`Most direct dependencies:`.padEnd(padding)}"${pa.name}": ${pa.directDependencyCount}`
    );
}

function printMostReferred(arg: [string, number], padding: number): void {
    console.log(`${`Most referred package:`.padEnd(padding)}"${arg[0]}": ${arg[1]}x`);
}

function printMostVersion(mostVerions: VersionSummary, padding: number, paddingLeft: number): void {
    console.log(`${`Package(s) with most versions:`.padEnd(padding)}`);

    for (const [name, versions] of mostVerions) {
        console.log(`${``.padStart(paddingLeft)}${name} [${[...versions].join(", ")}]`);
    }
}

function printLicenseInfo(groupedLicenses: GroupedLicenseSummary, paddingLeft: number): void {
    console.log(`Licenses:`);
    for (const { license, names } of groupedLicenses) {
        const threshold = 4;
        const samples: string[] = names.slice(0, threshold);

        if (names.length >= threshold)
            console.log(
                `${``.padStart(paddingLeft)}${license} - [${samples.join(", ")}, +${
                    names.length - threshold
                } more]`
            );
        else console.log(`${``.padStart(paddingLeft)}${license} - [${samples.join(", ")}]`);
    }
}
