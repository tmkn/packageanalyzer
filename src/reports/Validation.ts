import { z } from "zod";

export const BasePackageParameter = z.object({
    package: z.string()
});

export const BaseFolderParameter = z.object({
    folder: z.string()
});

export const dependencyTypes = z.union([z.literal(`dependencies`), z.literal(`devDependencies`)], {
    invalid_type_error: `type must be "dependencies" or "devDependencies"`
});

export type DependencyTypes = z.infer<typeof dependencyTypes>;

export const TypeParameter = z.object({
    type: dependencyTypes
});
