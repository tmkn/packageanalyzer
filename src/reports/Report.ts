import { isRight } from "fp-ts/lib/Either";
import * as t from "io-ts";

import { IDecorator } from "../extensions/decorators/Decorator";
import { Package } from "../package/package";
import { IPackageJsonProvider } from "../providers/provider";
import { IFormatter } from "../utils/formatter";
import { PackageVersion } from "../visitors/visitor";
import { DependencyTypes } from "./Validation";

export interface IReportContext {
    stdoutFormatter: IFormatter;
    stderrFormatter: IFormatter;
}

export interface IReport<T extends {}> {
    readonly name: string;
    readonly params: T | undefined;
    readonly pkg: PackageVersion;

    readonly decorators?: IDecorator<any, any>[];
    readonly provider?: IPackageJsonProvider;
    readonly type?: DependencyTypes;
    readonly depth?: number;

    report(pkg: Package, context: IReportContext): Promise<void>;
    validate?(): t.Type<T>;
}

export abstract class AbstractReport<T extends {}> implements IReport<T> {
    abstract name: string;
    readonly params: T | undefined = undefined;
    abstract pkg: PackageVersion;

    decorators: IDecorator<any, any>[] | undefined;
    provider: IPackageJsonProvider | undefined;
    type: DependencyTypes | undefined;
    depth: number | undefined;

    constructor(params: T) {
        const foo = this.validate?.();

        if (foo) {
            const data = foo.decode(params);

            if (isRight(data)) {
                this.params = data.right;
            }
        }
    }

    abstract report(pkg: Package, context: IReportContext): Promise<void>;
    abstract validate?(): t.Type<T, T, unknown>;
}
