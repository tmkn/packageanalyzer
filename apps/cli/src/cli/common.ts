import { Command } from "clipanion";

import { AbstractReport } from "../../../../packages/shared/src/reports/Report.js";
import { ReportService } from "../../../../packages/shared/src/reports/ReportService.js";
import { Formatter, type IFormatter } from "../../../../packages/shared/src/utils/formatter.js";
import { NodeHost } from "../../../../packages/node/src/host/NodeHost.js";

export abstract class CliCommand<
    T extends AbstractReport<any> | AbstractReport<any>[]
> extends Command {
    abstract getReports(): T | Promise<T>;

    exitCode = 0;

    beforeProcess: ((report: T) => void) | undefined = undefined;

    async execute(): Promise<number | void> {
        const host = new NodeHost(this.context.stdout, this.context.stderr);

        try {
            const reports = await this.getReports();
            const reportService = new ReportService(
                {
                    reports: Array.isArray(reports) ? reports : [reports]
                },
                host
            );

            this.beforeProcess?.(reports);
            this.exitCode = (await reportService.process()) ?? 0;
        } catch (e: unknown) {
            const stderrFormatter: IFormatter = new Formatter(host.getStderrWriter());

            if (e) stderrFormatter.writeLine(e?.toString());
            this.exitCode = 1;
        } finally {
            await Promise.all([host.getStdoutWriter().flush(), host.getStderrWriter().flush()]);
        }

        return this.exitCode;
    }
}
