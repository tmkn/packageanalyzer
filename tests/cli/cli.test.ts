import { describe, test, expect } from "vitest";

import { CliCommand } from "../../src/cli/common.js";
import { AbstractReport } from "../../src/reports/Report.js";
import { isValidDependencyType } from "../../src/reports/Validation.js";
import { createMockContext, TestReport } from "../common.js";
import { MockProvider, type IMockPackageJson } from "../mocks.js";

describe(`CLI Utility`, () => {
    test(`isValidDependencyType`, () => {
        expect(isValidDependencyType("dependencies")).toEqual(true);
        expect(isValidDependencyType("devDependencies")).toEqual(true);
        expect(isValidDependencyType("abc")).toEqual(false);
        expect(isValidDependencyType(3)).toEqual(false);
    });
});

describe(`CliCommand Tests`, () => {
    test(`Correctly writes to stderr on exception`, async () => {
        const { command, stdout, stderr } = createMockCliCommand(() => {
            throw new Error(`Whoops`);
        });

        await command.execute();

        expect(stdout.lines).toMatchSnapshot(`stdout`);
        expect(stderr.lines).toMatchSnapshot(`stderr`);
    });

    describe(`getReports`, () => {
        test(`returns a single report`, async () => {
            expect.assertions(3);

            const { command } = createMockCliCommand(() => {
                const report = createReport(async packages => {
                    expect(packages).toHaveLength(1);
                    expect(packages[0].name).toEqual(packageJson.name);
                    expect(packages[0].version).toEqual(packageJson.version);
                });

                return report;
            });

            await command.execute();
        });

        test(`async returns a single report`, async () => {
            expect.assertions(3);

            const { command } = createMockCliCommand(async () => {
                const report = createReport(async packages => {
                    expect(packages).toHaveLength(1);
                    expect(packages[0].name).toEqual(packageJson.name);
                    expect(packages[0].version).toEqual(packageJson.version);
                });

                return report;
            });

            await command.execute();
        });

        test(`returns multiple reports`, async () => {
            expect.assertions(6);

            const { command } = createMockCliCommand(() => {
                const report1 = createReport(async packages => {
                    expect(packages).toHaveLength(1);
                    expect(packages[0].name).toEqual(packageJson.name);
                    expect(packages[0].version).toEqual(packageJson.version);
                });

                const report2 = createReport(async packages => {
                    expect(packages).toHaveLength(1);
                    expect(packages[0].name).toEqual(packageJson.name);
                    expect(packages[0].version).toEqual(packageJson.version);
                });

                return [report1, report2];
            });

            await command.execute();
        });

        test(`async returns multiple reports`, async () => {
            expect.assertions(6);

            const { command } = createMockCliCommand(async () => {
                const report1 = createReport(async packages => {
                    expect(packages).toHaveLength(1);
                    expect(packages[0].name).toEqual(packageJson.name);
                    expect(packages[0].version).toEqual(packageJson.version);
                });

                const report2 = createReport(async packages => {
                    expect(packages).toHaveLength(1);
                    expect(packages[0].name).toEqual(packageJson.name);
                    expect(packages[0].version).toEqual(packageJson.version);
                });

                return [report1, report2];
            });

            await command.execute();
        });
    });
});

type ReportReturnType = AbstractReport<any> | AbstractReport<any>[];

interface MockCliCommandParams {
    getReports?: () => ReportReturnType | Promise<ReportReturnType>;
}

class MockCliCommand extends CliCommand<ReportReturnType> {
    constructor(private _params: MockCliCommandParams) {
        super();

        if (this._params.getReports) {
            this.getReports = this._params.getReports;
        }
    }

    getReports(): ReportReturnType | Promise<ReportReturnType> {
        throw new Error(`getReports method not implemented`);
    }
}

const packageJson: IMockPackageJson = { name: `mockpackage`, version: `1.2.3` };
const provider: MockProvider = new MockProvider([packageJson]);

function createMockCliCommand(getReports: MockCliCommandParams["getReports"]) {
    const command = new MockCliCommand({
        getReports
    });
    const { mockContext, ...rest } = createMockContext();
    command.context = mockContext;

    return { command, ...rest };
}

function createReport(reportFn: AbstractReport<any>["report"]): AbstractReport<any> {
    const report = new TestReport({
        pkg: [packageJson.name!, packageJson.version],
        report: async (...args) => reportFn(...args)
    });

    report.provider = provider;

    return report;
}
