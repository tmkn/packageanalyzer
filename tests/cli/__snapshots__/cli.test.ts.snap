// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CliCommand Tests Correctly writes to stderr on exception: stderr 1`] = `
[
  "Error: Whoops",
  "",
]
`;

exports[`CliCommand Tests Correctly writes to stderr on exception: stdout 1`] = `[]`;
